File: cookieManager.js
// cookieManager.js

require('dotenv').config(); // Ensure .env is loaded

let cookies = process.env.DEEPL_COOKIES ? process.env.DEEPL_COOKIES.split(',') : [];
let invalidCookies = [];
let currentCookieIndex = 0;

// Validate cookies format
function validateCookies() {
  cookies = cookies.filter(cookie => /^[0-9a-fA-F-]{8}-[0-9a-fA-F-]{4}-[0-9a-fA-F-]{4}-[0-9a-fA-F-]{4}-[0-9a-fA-F-]{12}$/.test(cookie));
  if (cookies.length === 0) {
    console.error("No valid cookies provided. Please check your DEEPL_COOKIES environment variable.");
    console.error("Cookies found:", process.env.DEEPL_COOKIES);
    process.exit(1);
  }
}

function getNextCookie() {
  let attempts = 0;
  while (attempts < cookies.length) {
    const cookie = cookies[currentCookieIndex];
    if (!invalidCookies.includes(cookie)) {
      currentCookieIndex = (currentCookieIndex + 1) % cookies.length;
      return `dl_session=${cookie}`;
    }
    currentCookieIndex = (currentCookieIndex + 1) % cookies.length;
    attempts++;
  }
  return null;
}

function markCookieInvalid(cookie) {
  const cookieValue = cookie.replace('dl_session=', '');
  if (!invalidCookies.includes(cookieValue)) {
    invalidCookies.push(cookieValue);
  }
}

// Initialize and validate cookies on startup
validateCookies();

module.exports = {
  getNextCookie,
  markCookieInvalid
};

File: translator/cookie_manager.go
package translator

import (
	"fmt"
	"log"
	"os"
	"strings"
)

var cookies []string
var invalidCookies []string
var currentCookieIndex int

func validateCookies() {
	cookieStr := os.Getenv("DEEPL_COOKIES")
	if cookieStr == "" {
		log.Fatal("No cookies provided. Please check your DEEPL_COOKIES environment variable.")
	}
	cookieList := strings.Split(cookieStr, ",")
	for _, cookie := range cookieList {
		cookie = strings.TrimSpace(cookie)
		if len(cookie) == 36 {
			cookies = append(cookies, cookie)
		} else {
			log.Printf("Invalid cookie format: %s", cookie)
		}
	}
	if len(cookies) == 0 {
		log.Fatal("No valid cookies provided. Please check your DEEPL_COOKIES environment variable.")
	}
}

func getNextCookie() (string, error) {
	attempts := 0
	for attempts < len(cookies) {
		cookie := cookies[currentCookieIndex]
		if !stringSliceContains(invalidCookies, cookie) {
			currentCookieIndex = (currentCookieIndex + 1) % len(cookies)
			return cookie, nil
		}
		currentCookieIndex = (currentCookieIndex + 1) % len(cookies)
		attempts++
	}
	return "", fmt.Errorf("no valid cookies available")
}

func markCookieInvalid(cookie string) {
	if !stringSliceContains(invalidCookies, cookie) {
		invalidCookies = append(invalidCookies, cookie)
	}
}

func initCookies() {
	validateCookies()
}
